modules/dreamview/backend/perception_camera_updater/perception_camera_updater.cc

```diff
@@ -0,0 +1,221 @@
+ *****************************************************************************/
+
+#include "modules/dreamview/backend/perception_camera_updater/perception_camera_updater.h"
+
+#include <float.h>
+#include <map>
+#include <string>
+#include "cyber/common/file.h"
+#include "modules/common/adapters/adapter_gflags.h"
+#include "opencv2/opencv.hpp"
+
+namespace apollo {
+namespace dreamview {
+
+using apollo::drivers::CompressedImage;
+using apollo::localization::LocalizationEstimate;
+using apollo::localization::Pose;
+using apollo::transform::TransformStampeds;
+using apollo::transform::Transform;
+
+namespace {
+  void ConvertMatrixToArray(
+      const Eigen::Matrix4d &matrix, std::vector<double> *array) {
+    const double* pointer = matrix.data();
+    for (int i = 0; i < matrix.size(); ++i) {
+      array->push_back(pointer[i]);
+    }
+  }
+}  // namespace
+
+PerceptionCameraUpdater::PerceptionCameraUpdater(WebSocketHandler* websocket)
+    : websocket_(websocket),
+      node_(cyber::CreateNode("perception_camera_updater")) {
+  InitReaders();
+}
+
+void PerceptionCameraUpdater::Start() {
+  enabled_ = true;
+}
+
+void PerceptionCameraUpdater::Stop() {
+  if (enabled_) {
+    localization_queue_.clear();
+    image_buffer_.clear();
+    tf_static_.clear();
+    current_image_timestamp_ = 0.0;
+  }
+  enabled_ = false;
+}
+
+void PerceptionCameraUpdater::GetImageLocalization(
+    std::vector<double> *localization) {
+  double timestamp_diff = DBL_MAX;
+  auto iter = localization_queue_.rbegin();
+  Pose image_pos;
+  for (; iter != localization_queue_.rend(); ++iter) {
+    const double tmp_diff = (*iter)->measurement_time()
+        - current_image_timestamp_;
+    if (tmp_diff > 0) {
+      timestamp_diff = tmp_diff;
+      image_pos = (*iter)->pose();
+    } else {
+      if (std::abs(tmp_diff) < timestamp_diff) {
+        image_pos = (*iter)->pose();
+      }
+      break;
+    }
+  }
+  // clean useless localization
+  auto iter_begin = localization_queue_.begin();
+  while ((*iter_begin) != (*iter)) {
+    iter_begin = localization_queue_.erase(iter_begin);
+  }
+
+  Eigen::Matrix4d localization_matrix;
+  localization_matrix.setConstant(0);
+  Eigen::Quaterniond q;
+  q.x() = image_pos.orientation().qx();
+  q.y() = image_pos.orientation().qy();
+  q.z() = image_pos.orientation().qz();
+  q.w() = image_pos.orientation().qw();
+  localization_matrix.block<3, 3>(0, 0) = q.normalized().toRotationMatrix();
+  localization_matrix(0, 3) = image_pos.position().x();
+  localization_matrix(1, 3) = image_pos.position().y();
+  localization_matrix(2, 3) = image_pos.position().z();
+  localization_matrix(3, 3) = 1;
+
+  ConvertMatrixToArray(localization_matrix, localization);
+}
+
+void PerceptionCameraUpdater::OnImage(
+    const std::shared_ptr<CompressedImage> &compressed_image) {
+  if (!enabled_
+      || compressed_image->format() == "h265" /* skip video format */) {
+    return;
+  }
+
+  std::lock_guard<std::mutex> lock(image_mutex_);
```

Comment:
is lock still required after line 117?

Reply:
Lock is required to save image timestamp and image itself, I moved the lock behind to exclude the image compression process.